### PR TITLE
Add check for C++17 std::transform_reduce

### LINKED ISSUE #1617: Insufficient C++17 compiler support check in CMake
**Please describe the problem you are trying to solve.**

Using GCC 9.2.0 and our latest develop, I am able to configure precice using `CMake`. However, at compile time, the compiler generates the following error message:

```bash
~/precice/src/com/SerializedPartitioning.cpp: In static member function ‘static precice::com::serialize::SerializedConnectionMap precice::com::serialize::SerializedConnectionMap::serialize(const ConnectionMap&)’:
~/precice/src/com/SerializedPartitioning.cpp:26:8: error: no matching function for call to ‘transform_reduce(std::map<int, std::vector<int> >::const_iterator, std::map<int, std::vector<int> >::const_iterator, int, std::plus<long unsigned int>, precice::com::serialize::SerializedConnectionMap::serialize(const ConnectionMap&)::<lambda(const auto:1&)>)’
   26 |       });
      |        ^
```

The problem here is that this GCC version does not fully support the required [`std::transform_reduce`](https://en.cppreference.com/w/cpp/algorithm/transform_reduce) C++17 feature.

**Describe the solution you propose.**

We should catch this already at compile time and inform the user about this. Looking at our CMake C++ checks, our standard support check [only checks for C++11 compliance](https://github.com/precice/precice/blob/develop/cmake/CheckCXX11Library.cmake). It would be easy to either extend the check to C++17 or add an additional C++17 check.

For me, it was sufficient to modify the C++11 check for 

```cpp
#include <execution>
```

**Describe alternatives you've considered**
Do nothing and let the user figure out.

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR adds a check for C++17 std::transform_reduce during the configuration of the CMake project.

## Motivation and additional information

Closes #1617

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.16.3.
* [x] I squashed / am about to squash all commits that should be seen as one.

### CODE REVIEW

File: cmake/CheckSTL.cmake [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  set(CMAKE_CXX_STANDARD 17)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  check_cxx_source_compiles("${CheckSourceTransformReduce}" CXX17_TRANSFORM_REDUCE)
+  if(NOT "${CXX17_TRANSFORM_REDUCE}")
+    message(FATAL_ERROR "
+    The standard library used by your compiler ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION} is not C++17 compliant.
+    It does not support std::transform_reduce (P0024R2)!
+    Please upgrade your compiler at least to:
+    * GCC 9.2
```
Comment (davidscn) [inline cmake/CheckSTL.cmake:50]: I used this GCC version when facing the issue in the first place. I guess this should be 9.3?

File: cmake/CheckSTL.cmake [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  int main() { std::time_t t = std::time(nullptr); std::tm tm = *std::localtime(&t); std::cout << std::put_time(&tm, \"%c %Z\"); }"
+  )
+
+  set(CMAKE_CXX_STANDARD 11)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  check_cxx_source_compiles("${CheckSourceN2071N2072}" CXX11_N2071_N2072)
+  if(NOT "${CXX11_N2071_N2072}")
+    message(FATAL_ERROR "
+    The standard library used by your compiler ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION} is not C++11 compliant.
+    It does not support time manipulation (N2071 and N2072)!
```
Comment (davidscn) [inline cmake/CheckSTL.cmake:18]:
```suggestion
    It does not support time manipulation (N2071 and N2072).
```

File: cmake/CheckSTL.cmake [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    Please upgrade your compiler at least to:
+    * GCC 5
```
Comment (davidscn) [inline cmake/CheckSTL.cmake:20]: I guess this is not really of high practical relevance, as almost all compilers will succeed here, but it is somewhat weird to ask for GCC v5, as it won't be sufficient to compile preCICE.

File: cmake/CheckSTL.cmake (hunk 1 above)
Comment (fsimonis) [reply cmake/CheckSTL.cmake:50]:
```suggestion
    * GCC 9.3
```

File: cmake/CheckSTL.cmake (hunk 3 above)
Comment (fsimonis) [reply cmake/CheckSTL.cmake:20]: True. We should provide a global recommended baseline for compilers.

File: cmake/CheckSTL.cmake [hunk 4]
```diff
@@ -0,0 +1,71 @@
+include(CheckCXXSourceCompiles)
+
+# These are the compiler/STL versions required to compile preCICE
+set(_precice_recommended_baseline "Please upgrade your compiler at least to:
+    * GCC 9.2
+    * LLVM 5.0
+    * MSVC 19.14
+    * Intel 15 in conjunction with GCC 9.2
+    * PGI 2015 in conjunction with GCC 9.2
```
Comment (davidscn) [inline cmake/CheckSTL.cmake:9]:
```suggestion
    * GCC 9.3
    * LLVM 5.0
    * MSVC 19.14
    * Intel 15 in conjunction with GCC 9.3
    * PGI 2015 in conjunction with GCC 9.3
```

**Pull request merged**